core/compiler.common/src/org/jetbrains/kotlin/types/AbstractTypeChecker.kt

```diff
@@ -407,9 +407,10 @@ object AbstractTypeChecker {
         val superTypeCaptured = superType.asCapturedType()
         val lowerType = superTypeCaptured?.lowerType()
         if (superTypeCaptured != null && lowerType != null) {
+            val properSubType = if (subType.isMarkedNullable() && superType.isMarkedNullable()) subType.withNullability(false) else subType
```

Comment:
How about adding some comments around here to explain the case? Like,

// If both are nullable, e.g., `T? <: Captured<in T>?`, we want non-null variant of LHS, `T`, to be compared
// against lower captured type of RHS, `Captured<in T>`.

if my understanding is correct.

Also the commit message can be more descriptive with a similar explanation.

Reply:
done

Reply:
It's not clear to me why we should modify **subType** and not superType. Basically, here we're trying to represent (approximate) type `Captured(in Foo)?` somehow, so why not add nullability for lower type? There are some implications if `Foo` is a flexible type (we can add constraint `T <: Foo!` or `T <: Foo?`)

Reply:
By the way, shouldn't we modify types when `subType` is not-nullable? Consider a situation like `T <: Captured(in Foo)?` where `T` is a type variable. Currently, we'll end up adding constraint like `T <: Foo` and, as a result, we'll infer `T` to `Foo`, while it should be inferred to `Foo?`

Reply:
Updated the PR to use nullable version of the lower type if the captured type is marked nullable. Also updated the names.

Reply:
Sorry for the confusion but one more addition: let's remove `lowerType.isSimpleType()` as it seems we should perform the same approximation for kinds of lower types